### PR TITLE
ARROW-4087: [C++] Make CSV spellings of null values configurable

### PR DESCRIPTION
**Pull request by pitrou**
Interestingly, there is no noticeable slowdown when reading CSV files (even though the trie is significantly slower than the hard-coded function in microbenchmarks).

### CODE REVIEW

**Comment by bkietz** [comment]
Is the trie faster than a flat hash table?

**Comment by pitrou** [comment]
I haven't compared directly, but we can compute the hash of approximately 250M small strings/second, which is an upper bound for a full hash table lookup. The trie is faster than that when the string is not in the trie... My intuition is that the hash table is unlikely to be as performant for this use case (a small number of short strings).

**Comment by fsaintjacques** [comment]
Small vector and linear scan should be the fastest when N is very small, which it is here. Bonus point if you can make the allocation on the cacheline after the vector struct.

**Comment by pitrou** [comment]
Hmm, I hadn't thought about a linear scan.

**Comment by fsaintjacques** [comment]
And with small string optimization (which we have a good chance of hitting), everything will be on the stack of the function and on the same "near" cachelines without pointer chasing.

**Comment by wesm** [comment]
Aside: we need to curate that nulls list because NaN is not null in Arrow-land.

**Comment by wesm** [comment]
We should look at bloom filters at some point, too (there is an implementation of one under src/parquet already, albeit with virtual calls on the inner loop). They are popular for these "not contained in" checks, but I don't know where the crossover point in table size where a bloom filter becomes worthwhile

**Comment by pitrou** [comment]
AFAIK, bloom filters are useful when your table is too large to fit in cache. The addressed use case here is a small table.

(though, CPython uses a trivial bloom filter to speed up substring searches...)

File: cpp/src/arrow/util/trie-test.cc [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+TEST(Trie, Basics3) {
+  TestTrieContents({"abcd", "ab", "a"});
+  TestTrieContents({"abcd", "ab", "a", ""});
+}
+
+TEST(Trie, LongStrings) {
+  TestTrieContents({"abcdefghijklmnopqr", "abcdefghijklmnoprq", "defghijklmnopqrst"});
+  TestTrieContents({"abcdefghijklmnopqr", "abcdefghijklmnoprq", "abcde"});
```
Comment (wesm) [inline cpp/src/arrow/util/trie-test.cc]: The long strings only create 2 nodes (right?), not sure if there's value in having to create more than 2 to ensure that the recursion is happening properly

**Review by wesm — APPROVED** [review]
+1 -- the last time I encountered the trie data structure was years ago in algorithms class. It's nice to see it pop up in a practical setting. This implementation is nice -- you might consider writing a brief blog post about it in the Arrow blog

**Comment by wesm** [comment]
On my hardware (xeon e3-1505m @ 3.7GHz) the Trie beats the inlined in not found

```
$ ./release/arrow-trie-benchmark 
2018-12-20 16:21:37
Running ./release/arrow-trie-benchmark
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
BM_TrieLookupFound/repeats:2                         635 ns        635 ns    1093358   150.116M items/s
BM_TrieLookupFound/repeats:2                         636 ns        636 ns    1093358   149.879M items/s
BM_TrieLookupFound/repeats:2_mean                    636 ns        636 ns    1093358   149.997M items/s
BM_TrieLookupFound/repeats:2_median                  636 ns        636 ns    1093358   149.997M items/s
BM_TrieLookupFound/repeats:2_stddev                    1 ns          1 ns    1093358   171.131k items/s
BM_TrieLookupNotFound/repeats:2                      234 ns        234 ns    3014197   407.312M items/s
BM_TrieLookupNotFound/repeats:2                      235 ns        235 ns    3014197   405.283M items/s
BM_TrieLookupNotFound/repeats:2_mean                 235 ns        235 ns    3014197   406.297M items/s
BM_TrieLookupNotFound/repeats:2_median               235 ns        235 ns    3014197   406.297M items/s
BM_TrieLookupNotFound/repeats:2_stddev                 1 ns          1 ns    3014197   1.43421M items/s
BM_InlinedTrieLookupFound/repeats:2                  269 ns        269 ns    2578567   354.334M items/s
BM_InlinedTrieLookupFound/repeats:2                  268 ns        268 ns    2578567   355.394M items/s
BM_InlinedTrieLookupFound/repeats:2_mean             269 ns        269 ns    2578567   354.864M items/s
BM_InlinedTrieLookupFound/repeats:2_median           269 ns        269 ns    2578567   354.864M items/s
BM_InlinedTrieLookupFound/repeats:2_stddev             1 ns          1 ns    2578567   767.929k items/s
BM_InlinedTrieLookupNotFound/repeats:2               243 ns        243 ns    2879162   392.837M items/s
BM_InlinedTrieLookupNotFound/repeats:2               243 ns        243 ns    2879162   393.025M items/s
BM_InlinedTrieLookupNotFound/repeats:2_mean          243 ns        243 ns    2879162   392.931M items/s
BM_InlinedTrieLookupNotFound/repeats:2_median        243 ns        243 ns    2879162   392.931M items/s
BM_InlinedTrieLookupNotFound/repeats:2_stddev          0 ns          0 ns    2879162   136.835k items/s
```

File: cpp/src/arrow/util/trie-test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/trie-test.cc]: Hmm... right, our small strings should be 11 chars long.

**Comment by pitrou** [comment]
For the record, I tried replacing the trie with a hash table: it makes CSV parsing (for columns of integers) around 10% slower.

**Pull request closed without merging**